source/common/stats/hystrix.cc

```diff
@@ -0,0 +1,319 @@
+#include "common/stats/hystrix.h"
+
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <sstream>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/logger.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Stats {
+
+const uint64_t Hystrix::DEFAULT_NUM_OF_BUCKETS;
+
+// Add new value to rolling window, in place of oldest one.
+void Hystrix::pushNewValue(const std::string& key, uint64_t value) {
+  // Create vector if do not exist.
+  // TODO trabetti: why resize + value param didn't work without the if?
+  if (rolling_stats_map_.find(key) == rolling_stats_map_.end()) {
+    rolling_stats_map_[key].resize(num_of_buckets_, value);
+  } else {
+    rolling_stats_map_[key][current_index_] = value;
+  }
+}
+
+uint64_t Hystrix::getRollingValue(absl::string_view cluster_name, absl::string_view stats) {
+  std::string key;
+  key = absl::StrCat("cluster.", cluster_name, ".", stats);
+  if (rolling_stats_map_.find(key) != rolling_stats_map_.end()) {
```

Comment:
```
auto iter = rolling_stats_map_.find(key);
if (iter != rolling_stats_map_.end()) {
  RollingStats& rolling_stats =  iter->second;
  if (rolling_stats[current_index] < rolling_stats[(current_index_ + 1) % num_buckets_])
....

Try to do the map lookups only once on the same key within a function.

Reply:
Before the sink implementation, we had multiple variables to pass to the handler, so it made sense to add something generic like the `handlerInfo`, to pass all of them in one class. We also thought that other handlers may be able to use it, instead of adding more and more arguments.
Now that we moved to the sink implementation, and we only need the callbacks from the filter,  one option is to get rid of the `handlerInfo` and add instead the callbacks as an arguments to the handler interface.
Another direction - We also talked about changing the handlers from methods to classes (in a separate issue and a separate PR). I didn't think this through, but I this may also solve the problem.